plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol

```diff
@@ -0,0 +1,224 @@
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+import "../PaymentExitDataModel.sol";
+import "../PaymentInFlightExitModelUtils.sol";
+import "../../../vaults/EthVault.sol";
+import "../../../vaults/Erc20Vault.sol";
+import "../../../framework/PlasmaFramework.sol";
+
+library PaymentProcessInFlightExit {
+    using PaymentInFlightExitModelUtils for PaymentExitDataModel.InFlightExit;
+
+    uint8 public constant MAX_INPUT_NUM = 4;
+    uint8 public constant MAX_OUTPUT_NUM = 4;
+
+    struct Controller {
+        PlasmaFramework framework;
+        EthVault ethVault;
+        Erc20Vault erc20Vault;
+    }
+
+    event InFlightExitOmitted(
+        uint192 indexed exitId,
+        address token
+    );
+
+    event InFlightExitOutputWithdrawn(
+        uint192 indexed exitId,
+        uint16 outputIndex
+    );
+
+    event InFlightExitInputWithdrawn(
+        uint192 indexed exitId,
+        uint16 inputIndex
+    );
+
+    function run(
+        Controller memory self,
+        PaymentExitDataModel.InFlightExitMap storage exitMap,
+        uint192 exitId,
+        address token
+    )
+        public
+    {
+        PaymentExitDataModel.InFlightExit storage exit = exitMap.exits[exitId];
+
+        // check if any input spent already, this is required to prevent operator stealing fund.
+        // Since process exit should not revert to avoid blocking the while loop, return directly.
+        // see: https://github.com/omisego/plasma-contracts/issues/102#issuecomment-495809967
+        if (exit.exitStartTimestamp == 0 || isAnyInputSpent(self.framework, exit, token)) {
+            emit InFlightExitOmitted(exitId, token);
+            return;
+        }
+
+        if (!exit.isCanonical) {
+            for (uint16 i = 0; i < MAX_INPUT_NUM; i++) {
+                PaymentExitDataModel.WithdrawData memory withdrawal = exit.inputs[i];
+
+                if (shouldWithdrawInput(exit, withdrawal, token, i)) {
+                    withdrawFromVault(self, withdrawal);
+                    withdrawal.exitTarget.transfer(withdrawal.piggybackBondSize);
+                    emit InFlightExitInputWithdrawn(exitId, i);
+                }
+            }
+        } else {
+            for (uint16 i = 0; i < MAX_OUTPUT_NUM; i++) {
+                PaymentExitDataModel.WithdrawData memory withdrawal = exit.outputs[i];
+
+                if (shouldWithdrawOutput(self, exit, withdrawal, token, i)) {
+                    withdrawFromVault(self, withdrawal);
+                    withdrawal.exitTarget.transfer(withdrawal.piggybackBondSize);
+                    emit InFlightExitOutputWithdrawn(exitId, i);
+                }
+            }
+        }
+
+        flagInputAndOutputSpent(self.framework, exit, token);
```

Comment:
I would like to keep `flagInputAndOutputSpent` to process together to optimize gas a bit (calls the `framework.batchFlagOutputsSpent` once instead of two) and I think the readability/simplicity improvement is fairly small in this case.